cpp/src/community/leiden_kernels.cu

```diff
@@ -0,0 +1,302 @@
+  vertex_t const *d_dst_indices    = graph.indices;
+
+  weight_t new_Q = modularity<vertex_t, edge_t, weight_t>(
+    total_edge_weight, resolution, graph, cluster.data().get(), stream);
+
+  weight_t cur_Q = new_Q - 1;
+
+  // To avoid the potential of having two vertices swap clusters
+  // we will only allow vertices to move up (true) or down (false)
+  // during each iteration of the loop
+  bool up_down = true;
+
+  while (new_Q > (cur_Q + 0.0001)) {
+    cur_Q = new_Q;
+
+    compute_delta_modularity(total_edge_weight,
+                             resolution,
+                             graph,
+                             src_indices,
+                             vertex_weights,
+                             cluster_weights,
+                             cluster,
+                             cluster_hash,
+                             delta_Q,
+                             old_cluster_sum,
+                             stream);
+
+    // Filter out positive delta_Q values for nodes not in the same constraint group
+    thrust::for_each(
+      rmm::exec_policy(stream)->on(stream),
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(graph.number_of_edges),
+      [d_src_indices, d_dst_indices, d_constraint, d_delta_Q] __device__(vertex_t i) {
+        vertex_t start_cluster = d_constraint[d_src_indices[i]];
+        vertex_t end_cluster   = d_constraint[d_dst_indices[i]];
+        if (start_cluster != end_cluster) d_delta_Q[i] = weight_t{0.0};
+      });
+
+    assign_nodes(graph,
+                 delta_Q,
+                 cluster_hash,
+                 src_indices,
+                 next_cluster,
+                 vertex_weights,
+                 cluster_weights,
+                 up_down,
+                 stream);
+
+    up_down = !up_down;
+
+    new_Q = modularity<vertex_t, edge_t, weight_t>(
+      total_edge_weight, resolution, graph, next_cluster.data().get(), stream);
+
+    if (new_Q > cur_Q) { thrust::copy(next_cluster.begin(), next_cluster.end(), cluster.begin()); }
+  }
+
+  return cur_Q;
+}
+
+template float update_clustering_by_delta_modularity_constrained(
+  float,
+  float,
+  GraphCSRView<int32_t, int32_t, float> const &,
+  rmm::device_vector<int32_t> const &,
+  rmm::device_vector<float> const &,
+  rmm::device_vector<float> &,
+  rmm::device_vector<int32_t> &,
+  rmm::device_vector<int32_t> &,
+  cudaStream_t);
+
+template double update_clustering_by_delta_modularity_constrained(
+  double,
+  double,
+  GraphCSRView<int32_t, int32_t, double> const &,
+  rmm::device_vector<int32_t> const &,
+  rmm::device_vector<double> const &,
+  rmm::device_vector<double> &,
+  rmm::device_vector<int32_t> &,
+  rmm::device_vector<int32_t> &,
+  cudaStream_t);
+
+template <typename vertex_t, typename edge_t, typename weight_t>
+void leiden(GraphCSRView<vertex_t, edge_t, weight_t> const &graph,
+            weight_t *final_modularity,
```

Comment:
`final_modularity` should be passed by reference, rather than a pointer to a non-array memory location. 

Perhaps even more readable would be to pass mutables and results like `final_modularity` as (part of a tuple of) return values.

Reply:
Changed.